File: JAVA/Gestionformation/src/main/java/org/example/models/Formation.java
package org.example.models;

public class Formation {

    private int idFormation;
    private String titre;
    private String description;
    private String type;
    private int duree;
    private String organisme;
    private String objectifs;
    private Integer idRh; // ID du RH qui a créé cette formation
    private double moyenneRating;
    public Formation() {
    }

    // Constructeur sans ID (pour INSERT)
    public Formation(String titre, String description, String type, int duree, String organisme, String objectifs, Integer idRh) {
        this.titre = titre;
        this.description = description;
        this.type = type;
        this.duree = duree;
        this.organisme = organisme;
        this.objectifs = objectifs;
        this.idRh = idRh;
    }

    //Constructeur avec ID
    public Formation(int idFormation, String titre, String description,
                     String type, int duree, String organisme, String objectifs, Integer idRh) {
        this.idFormation = idFormation;
        this.titre = titre;
        this.description = description;
        this.type = type;
        this.duree = duree;
        this.organisme = organisme;
        this.objectifs = objectifs;
        this.idRh = idRh;
    }

    //getters et setters

    public int getIdFormation() {
        return idFormation;
    }

    public void setIdFormation(int idFormation) {
        this.idFormation = idFormation;
    }

    public String getTitre() {
        return titre;
    }

    public void setTitre(String titre) {
        this.titre = titre;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }

    public String getType() {
        return type;
    }

    public void setType(String type) {
        this.type = type;
    }

    public int getDuree() {
        return duree;
    }

    public void setDuree(int duree) {
        this.duree = duree;
    }

    public String getOrganisme() {
        return organisme;
    }

    public void setOrganisme(String organisme) {
        this.organisme = organisme;
    }
    public String getObjectifs() {
        return objectifs;
    }

    public void setObjectifs(String objectifs) {
        this.objectifs = objectifs;
    }

    public Integer getIdRh() {
        return idRh;
    }

    public void setIdRh(Integer idRh) {
        this.idRh = idRh;
    }

    public double getMoyenneRating() { return moyenneRating; }
    public void setMoyenneRating(double moyenneRating) { this.moyenneRating = moyenneRating; }

    @Override
    public String toString() {
        return "Formation{" +
                "idFormation=" + idFormation +
                ", titre='" + titre + '\'' +
                ", description='" + description + '\'' +
                ", type='" + type + '\'' +
                ", duree=" + duree +
                ", organisme='" + organisme + '\'' +
                ", objectifs='" + objectifs + '\'' +
                ", idRh=" + idRh +
                '}';
    }
}

File: src/Entity/Formation/Formation.php
<?php

namespace App\Entity\Formation;

use App\Repository\Formation\FormationRepository;
use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;

#[ORM\Entity(repositoryClass: FormationRepository::class)]
#[ORM\Table(name: 'formation')]
#[ORM\HasLifecycleCallbacks]
class Formation
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(name: 'id_formation')]
    private ?int $id = null;

    #[ORM\Column(length: 255)]
    #[Assert\NotBlank(message: "Le titre de la formation est obligatoire.")]
    #[Assert\Length(
        min:4,
        max: 255,
        minMessage: "Le titre doit faire au moins {{ limit }} caractères."
    )]
    private ?string $titre = null;

    #[ORM\Column(type: Types::TEXT, nullable: true)]
    #[Assert\NotBlank(message: "Veuillez fournir une description de la formation.")]
    private ?string $description = null;

    #[ORM\Column(length: 50)]
    #[Assert\NotBlank(message: "Veuillez choisir une catégorie.")]
    private ?string $type = null;

    #[ORM\Column]
    #[Assert\NotBlank(message: "La durée est obligatoire.")]
    #[Assert\Positive(message: "La durée doit être un nombre positif.")]
    private ?int $duree = null;

    #[ORM\Column(length: 255, nullable: true)]
    #[Assert\NotBlank(message: "L'organisme formateur est obligatoire.")]
    private ?string $organisme = null;

    #[ORM\Column(type: Types::TEXT, nullable: true)]
    private ?string $objectifs = null;

    #[ORM\Column(name: 'image', length: 255, nullable: true)]
    private ?string $imageUrl = null;

    #[ORM\Column(name: 'id_rh')]
    private ?int $rhId = null;

    #[ORM\Column(name: 'created_at', type: Types::DATETIME_MUTABLE, nullable: true)]
    private ?\DateTimeInterface $createdAt = null;

    #[ORM\Column(name: 'updated_at', type: Types::DATETIME_MUTABLE, nullable: true)]
    private ?\DateTimeInterface $updatedAt = null;

    /** @var Collection<int, SessionFormation> */
    #[ORM\OneToMany(targetEntity: SessionFormation::class, mappedBy: 'formation')]
    private Collection $sessions;

    public function __construct()
    {
        $this->sessions = new ArrayCollection();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getTitre(): ?string
    {
        return $this->titre;
    }

    public function setTitre(string $titre): static
    {
        $this->titre = $titre;
        return $this;
    }

    public function getDescription(): ?string
    {
        return $this->description;
    }

    public function setDescription(?string $description): static
    {
        $this->description = $description;
        return $this;
    }

    public function getType(): ?string
    {
        return $this->type;
    }

    public function setType(string $type): static
    {
        $this->type = $type;
        return $this;
    }

    public function getDuree(): ?int
    {
        return $this->duree;
    }

    public function setDuree(int $duree): static
    {
        $this->duree = $duree;
        return $this;
    }

    public function getOrganisme(): ?string
    {
        return $this->organisme;
    }

    public function setOrganisme(?string $organisme): static
    {
        $this->organisme = $organisme;
        return $this;
    }

    public function getObjectifs(): ?string
    {
        return $this->objectifs;
    }

    public function setObjectifs(?string $objectifs): static
    {
        $this->objectifs = $objectifs;
        return $this;
    }

    public function getImageUrl(): ?string
    {
        return $this->imageUrl;
    }

    public function setImageUrl(?string $imageUrl): static
    {
        $this->imageUrl = $imageUrl;
        return $this;
    }

    public function getRhId(): ?int
    {
        return $this->rhId;
    }

    public function setRhId(int $rhId): static
    {
        $this->rhId = $rhId;
        return $this;
    }

    public function getCreatedAt(): ?\DateTimeInterface
    {
        return $this->createdAt;
    }

    public function getUpdatedAt(): ?\DateTimeInterface
    {
        return $this->updatedAt;
    }

    /** @return Collection<int, SessionFormation> */
    public function getSessions(): Collection
    {
        return $this->sessions;
    }

    #[ORM\PrePersist]
    public function onPrePersist(): void
    {
        $this->createdAt = new \DateTime();
        $this->updatedAt = new \DateTime();
    }

    #[ORM\PreUpdate]
    public function onPreUpdate(): void
    {
        $this->updatedAt = new \DateTime();
    }
}
